### PR TITLE
[WIP] Adding root models for local storage, implementation for secured vault

### PR DESCRIPTION
**Pull request by ikamthania**
### Do not merge this pull request yet as it is currently been worked upon.

**Objectives of this pull request are**

1. Setup application level root model which is a source of truth for the entire application.
2. Setup API to keep the root model in sync with local storage
3. Provide a vault mechanism where the sensitive data can be encrypted with the application password.

**What is done**
1. Root models are set up. The root model JSON which will be stored in the local storage is something similar to this 

```
{
	"ubunda": {
		"version": "v1.0"
	},
	"data": {
		"networks": [{
			"type": "ropsten",
			"rpcTarget": "https://ropsten.infura.io/metamask"
		}],
		"notices": [{
			"read": true,
			"date": "Thu Feb 09 2017",
			"title": "Terms of Use",
			"body": "",
			"id": 0
		}, {
			"read": true,
			"date": "Mon May 08 2017",
			"title": "Privacy Notice",
			"body": "",
			"id": 2
		}],
		"currencies": [{
			"currentCurrency": "usd",
			"conversionRate": 597.03,
			"conversionDate": 1521131153
		}],
		"keyrings": {
			"accounts": [{
				"publicKey": "0x02c55b67d11918e4cb165a8ef3f83ff8b2c9acadec62782d371b004e0e8f615cc3",
				"accountName": "Account 1"
			}, {
				"publicKey": "0x0330283f231c430e9a3054f1ef24b6ed0c43da2697554be442cafff2024aea9881",
				"accountName": "Account 2"
			}],
			"selectedAddress": "0x0330283f231c430e9a3054f1ef24b6ed0c43da2697554be442cafff2024aea9881",
			"vault": {
				"data": "nsIsghLe/Fqq8sRCMpR8ct/XAx/Oc6V97IB24CcwcYYNU47dmQyu7p9+HLGX0Ay7iBrKPyhjvNaVh6EyAxVIDiazQSj8hd906KnYg95YSSkn9D6ru+cypLNVmawN17MLD9prTij7Erhkt1/FcKafRUxguQrz8GlghypoMtAWcBN3RM4QmNvoOOHXcT/Am+1KhA==",
				"iv": "wE/Ro0wXZ0DyMq6yBUxQaQ==",
				"salt": "R/gdEqhZ1aCYHG4zFlHpDObI+QwOC/2iR44oU4ocgew="
			}
		},
		"preferencess": {
			"useBlockie": true,
			"featureFlags": [],
			"selectedTheme": "default",
			"selectedLanguage": "en_us"
		},
		"tokens": [],
		"infuraNetworkStatus": {
			"mainnet": "ok",
			"ropsten": "ok",
			"kovan": "ok",
			"rinkeby": "ok"
		}
	}
}
```
2. Secured vault is created which is encrypted with [browser-passworder](https://github.com/danfinlay/browser-passworder)
3. The content of the vault data is similar to this

```
{
	"mnemonic": "REDACTED",
	"hdDerivePath": "m/44'/60'/0'/0"
}
```
4. A subscription is added to application root model which listens to any change in the root model tree and updates the local storage as soon as their is a diff

```
def subscribeToAppRootChanges() = {
    updateRootModelFromStorage()
    WalletCircuit.subscribe(WalletCircuit.zoomTo(_.appRootModel)) { modelRO =>
      updateModelInLocalStorage()
    }
  }
```

The individual branches of this root model will be bind to different views of application

5. Separate handlers are added for different branches of root model and are then composed together

```
val appHandler = new AppHandler(zoomRW(_.appRootModel)((m, v) => m.copy(appRootModel = v)))
  val accountHandler = new KeyringHandler(zoomRW(_.appRootModel.appModel.data.keyrings)((m, v) =>
    m.copy(appRootModel = m.appRootModel
      .copy(appModel = m.appRootModel.appModel.copy(data = m.appRootModel.appModel.data.copy(keyrings = v))))))
  val appRootHandler = foldHandlers(appHandler, accountHandler)
```
6. Routes are separated into secured routes and open routes based on the condition 
```
.addCondition(getLoggedInStatus)(e => Some(redirectToPage(LandingLoc)(Redirect.Push)))
```

**What is in progress**

~~1. The backup view has number of moving pieces where at times it reads the model from props and sometimes from state, often in wrong manner, creating confusion how the data is moving. So the backup will be refactored so that instead of having three separate views it should be mounted conditionally based on state.~~

~~2. The api put together are in process of binding with the views, entire workflow binding is in progress in parallel.~~

~~3. The sign transaction has become a mix of statically typed scala.js code and dynamic javascript. Facades set up against the js code instead of libs, reducing the visibility of logic flow and opening room for runtime errors, that is being refactored so as to make it statically typed as far as possible~~

### CODE REVIEW

**Comment by Jake-Gillberg** [comment]
I think the 'keyring' datastructure should be more modular, not store information like "selected Address", and also support multiple vaults, if needed. Also, we probably want to store the [ethereum address, and not the public key](https://ethereum.stackexchange.com/questions/3542/how-are-ethereum-addresses-generated)

Maybe something like this:
```
"keyring": [
  {
    "addresses": [
      "0x02c55b67d11918e4cb165a8ef3f83ff8b2c9acadec62782d371b004e0e8f615cc3",
      "0x0330283f231c430e9a3054f1ef24b6ed0c43da2697554be442cafff2024aea9881"
    ],
    "vault": {
      "data": "nsIsghLe/Fqq8sRCMpR8ct/XAx/Oc6V97IB24CcwcYYNU47dmQyu7p9+HLGX0Ay7iBrKPyhjvNaVh6EyAxVIDiazQSj8hd906KnYg95YSSkn9D6ru+cypLNVmawN17MLD9prTij7Erhkt1/FcKafRUxguQrz8GlghypoMtAWcBN3RM4QmNvoOOHXcT/Am+1KhA==",
      "iv": "wE/Ro0wXZ0DyMq6yBUxQaQ==",
      "salt": "R/gdEqhZ1aCYHG4zFlHpDObI+QwOC/2iR44oU4ocgew="
    }
  }
]
```
We still need a place to tie Eth address (both those in "keyring" and those that aren't). Something like this:
```
"addressNameMap": [
  {
    "address": "0x02c55b67d11918e4cb165a8ef3f83ff8b2c9acadec62782d371b004e0e8f615cc3",
    "name": "Account 1"
  }
]
```
And someplace to store the
```
"activeAccount": "0x02c55b67d11918e4cb165a8ef3f83ff8b2c9acadec62782d371b004e0e8f615cc3"
```
As for ERC20 Tokens and MultiSig accounts, I imagine we want something like this:
```
"Contracts": {
  "ERC20s": [
    ERC20{}
  ],
  "MultiSigs": [
    MultiSig{}
  ]
}
```

**Comment by Jake-Gillberg** [comment]
As for cleaning up the transaction signing, what I would like to see is a parred down model where:
1. Simple validation of things like ethereum addresses being well formed or int out of range errors are done on type construction of the types in `Solidity.scala`
2. The model of how to get the raw transaction data is separate from the transaction signing: see `AddSharedWalletView.scala`
2.a EtherTransaction is a bad model and should be removed.

File: wallet-client/src/main/scala/com/livelygig/product/walletclient/facades/EthereumJsUtils.scala [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+import io.scalajs.nodejs.buffer.Buffer
+
+import scala.scalajs.js
+import scala.scalajs.js.annotation.JSImport
+
+@js.native
+@JSImport("ethereumjs-util", JSImport.Namespace)
+object EthereumJsUtils extends js.Object {
+  def bufferToHex(buffer: Buffer): String = js.native
```
Comment (Jake-Gillberg) [inline wallet-client/src/main/scala/com/livelygig/product/walletclient/facades/EthereumJsUtils.scala]: We shouldn't need a facade of this function, it should be included in the io.scalajs.nodejs.buffer.Buffer, see usage in https://github.com/LivelyGig/wallet/blob/f404f3d6063ba2f0ab569618984f32530056d0b0/wallet-client/src/main/scala/com/livelygig/product/walletclient/views/AddSharedWalletView.scala#L123

File: wallet-client/src/main/scala/com/livelygig/product/walletclient/rootmodel/AppRootModel.scala [hunk 2]
```diff
@@ -4,7 +4,6 @@ import com.livelygig.product.shared.models.wallet.AppModel
 
 case class AppRootModel(appModel: AppModel) {
   def updated(updatedAppModel: AppModel) = {
-    println("Yo I am in app root model")
```
Comment (Jake-Gillberg) [inline wallet-client/src/main/scala/com/livelygig/product/walletclient/rootmodel/AppRootModel.scala]: nice

File: wallet-client/src/main/scala/com/livelygig/product/walletclient/utils/TestApi.scala [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    val mnemonic = new Mnemonic()
+
+    val strongPassword = "my$trongP@ssword"
+
+    val mnemonicPhrase = mnemonic.toString
+
+    println(s"Root model created without vault ${Json.toJson(WalletCircuit.zoomTo(_.appRootModel.appModel).value)})")
+
+    println(s"Creating master private extended key with mnemonic phrase ${mnemonicPhrase}")
+
```
Comment (Jake-Gillberg) [inline wallet-client/src/main/scala/com/livelygig/product/walletclient/utils/TestApi.scala]: Is there stuff missing between print statements here? It seems like the file you were showing me on screenshare had more stuff.

File: wallet-client/src/main/scala/com/livelygig/product/walletclient/facades/EthereumJsUtils.scala (hunk 1 above)
Comment (ikamthania) [reply wallet-client/src/main/scala/com/livelygig/product/walletclient/facades/EthereumJsUtils.scala]: Will take a look thanks

File: wallet-client/src/main/scala/com/livelygig/product/walletclient/utils/TestApi.scala (hunk 3 above)
Comment (ikamthania) [reply wallet-client/src/main/scala/com/livelygig/product/walletclient/utils/TestApi.scala]: I guess no. I was, in fact, running the same version. Will be adding some more as I improve over things, just for a preview how APIs are going to look like.

**Comment by ikamthania** [comment]
Regarding the question about state and props. It's typical react js nomenclature. The state is associated with a component. Any change in the component should be tracked and modified by states. Props are properties which are defined on component creation. Typically, it's data that we need to pass to the component as the component is created. Changes I made in props is because there are few places where the state data is passed as props with a default value, which in my opinion is wrong. Any data with default value should be initiated in `initState` of the component rather than assuming that the data will be available at runtime.

File: wallet-client/src/main/scala/com/livelygig/product/walletclient/facades/EthereumJsUtils.scala (hunk 1 above)
Comment (ikamthania) [reply wallet-client/src/main/scala/com/livelygig/product/walletclient/facades/EthereumJsUtils.scala]: Noted. Will take a look, thanks.

**Comment by ikamthania** [comment]
@Jake-Gillberg regarding cleaning up the transaction signing, as discussed I will be picking it up and see to it how can that be improved over.

**Comment by ikamthania** [comment]
@Jake-Gillberg regarding keyring structure. I agree with your idea. I am wondering how the structures of `ERC20` and `MultiSig` are going to look like. There is a model for ERC20, but that needs some work for sure.

**Comment by Jake-Gillberg** [comment]
So, at their core, all that is really needed for the ERC20 and MultiSig data stores (and anything under `contracts`) is an ethereum `address`. But we might want to cache certain values from the contracts into the store like `decimals` for ERC20 or `owners` for a MultiSig wallet. (maybe we actually want a separate `contract-cache` data store?)

**Comment by ikamthania** [comment]
Ok. We will discuss it. Regarding the `contract-cache` I am assuming it will be persistent in local storage model and not in memory.

File: wallet-client/src/main/scala/com/livelygig/product/walletclient/utils/TestApi2.scala [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        try {
+          println(EthereumjsUnits.convert("0.01", "eth", "wei"))
+        } catch {
+          case e: Exception => e.printStackTrace()
+        }
+        println(res2)
+        try {
+          //          val myTransaction = new Transaction(new TransactionParams("0x4", "0x1e8f1c10800", "0x301114", s"0x${child2Address}", s"0x${EthereumjsUnits.convert("0.1", "eth", "wei")}", None, None, None, None))
+          //          val myTransaction = new Transaction("0x4", "0x1e8f1c10800", "0x301114", s"0x${child2Address}", s"0x${EthereumjsUnits.convert("0.1", "eth", "wei")}")
+          //todo the amount getting from EthereumjsUnits.convert("0.1", "eth", "wei") utility method is too high than actual amount.
```
Comment (Jake-Gillberg) [inline wallet-client/src/main/scala/com/livelygig/product/walletclient/utils/TestApi2.scala]: I seem to be getting the correct result of 100000000000000000 wei

Comment (ikamthania) [reply wallet-client/src/main/scala/com/livelygig/product/walletclient/utils/TestApi2.scala]: @Vinay-bhagwat can answer that. I think that error is from sending the raw transaction. So might be the case where there is an issue with the format of Transaction Params.

**Comment by ikamthania** [comment]
Little change in notices structure

```
"notices": [{
			"read": true,
			"date": "Thu Feb 09 2017",
			"title": "Terms of Use",
			"body": "",
			"type": "TermsOfServiceNotice"
		}, {
			"read": true,
			"date": "Mon May 08 2017",
			"title": "Privacy Notice",
			"body": "",
			"type": "PrivacyNotice"
		}]
```

Uses amazing [play-json-derived-codecs](https://github.com/julienrf/play-json-derived-codecs) for custom representation of Notice Types

**Comment by ikamthania** [comment]
Account Info moved out of keyring structure into separate accountInfo key

```
"accountInfo": [
  {
    "address": "0x02c55b67d11918e4cb165a8ef3f83ff8b2c9acadec62782d371b004e0e8f615cc3",
    "name": "Account 1"
  }
```
More data related to accounts can be added as required

**Comment by ikamthania** [comment]
Default account creation workflow is created. For it is assumed that the accounts will be derived from the mnemonic phrase. Will be adding workflow for importing accounts later.

**Comment by ikamthania** [comment]
Api calls need some work. It is scattered all over the place. Will consider typesafe rpc such as [autowire](https://github.com/lihaoyi/autowire) or [sloth](https://github.com/cornerman/sloth) or may be via 
play reverse routing https://github.com/jducoeur/Querki/blob/master/querki/scalajs/src/main/scala/querki/comm/ClientRoutes.scala

**Comment by ikamthania** [comment]
Most of the changes are already done. It has so many changes that reviewing is quite a task. Major functionality for account creation, secured vault, deriving child account, api for communicating with local storage are done. Will update the documentation. Will be merging this pull request with master in a while. Other changes can be done incrementally.

**Pull request merged**